### PR TITLE
use dynamic language setting for message time deltas

### LINKED ISSUE #2684: Dates are displayed in system locale regardless of language setting
Currently all the dates and times are diplayed regarding the android system locale using `DateUtils.formatDateTime`. 

Sometimes its even mixed because the plurals `hours_ago` and `minutes_ago` are translated.

My system language is set to german but I'm using TextSecure in english:
date

It would be more consistent if the dates are formatted regarding the textsecure language setting I guess.

### PR DESCRIPTION
**Pull request by rhodey**
modified DateUtils to use SimpleDateFormat for everything because unlike formatDateTime() it respects Locale.

Fixes #2684

### CODE REVIEW

File: src/org/thoughtcrime/securesms/util/DynamicLanguage.java [hunk 1]
```diff
@@ -49,7 +49,7 @@ private static Locale getActivityLocale(Activity activity) {
     return activity.getResources().getConfiguration().locale;
   }
 
-  private static Locale getSelectedLocale(Context context) {
+  public static Locale getSelectedLocale(Context context) {
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/util/DynamicLanguage.java]: return currentLocale instead

File: src/org/thoughtcrime/securesms/util/DateUtils.java [hunk 2]
```diff
@@ -39,6 +40,13 @@ private static int convertDelta(final long millis, TimeUnit to) {
     return (int) to.convert(System.currentTimeMillis() - millis, TimeUnit.MILLISECONDS);
   }
 
+  public static String getFormattedDateTime(Context context, long time, String template) {
+    final Locale locale           = DynamicLanguage.getSelectedLocale(context);
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/util/DateUtils.java]: Pass the locale in, rather than pulling it off a static DynamicLanguage method

**Comment by moxie0** [comment]
@mcginty will have to comment on the rest

File: src/org/thoughtcrime/securesms/util/DynamicLanguage.java (hunk 1 above)
Comment (rhodey) [inline src/org/thoughtcrime/securesms/util/DynamicLanguage.java]: can't modify this method like that, need to add a new non-static `getCurrentLocale()` method and then change the `DateUtils` method signatures to take a `DynamicLanguage` instance. doing this now but definite larger diff.

Comment (moxie0) [inline src/org/thoughtcrime/securesms/util/DynamicLanguage.java]: exactly. larger diff but not disgusting.

**Comment by moxie0** [comment]
whoa, this is way too much

**Comment by rhodey** [comment]
@moxie0 

+ `DateUtils` should use `DynamicLanguage`
+ `DynamicLanguage` is not static
+ `ConversationItem` and ConversationListItem` need to use `DateUtils`
+ `ConversationItem` and ConversationListItem` should probably use the same `DynamicLanguage` as their parent activities

So, lots of method signatures had to be changed to pass `DynamicLanguage` down. This diff can be drastically shrunk if `ConversationItem` and ConversationListItem` were allowed to create their own `DynamicLanguage` instance, else I don't feel this can be shrunk without refactoring `DynamicLanguage` to be static-- how about that?

**Comment by moxie0** [comment]
why not just pass a locale?

**Comment by rhodey** [comment]
@moxie0 ok, I agree that it makes more sense to pass in `Locale` instead of `DynamicLanguage`, but adding that change would not reduce the size of this diff because we'd just be passing down `DynamicLanguage.getCurrentLocale()` instead of `DynamicLanguage`.

**Comment by moxie0** [comment]
Why do you need a provider interface? Just pass the locale instead of the interface.

**Comment by rhodey** [comment]
@moxie0 `ConversationFragment` and `ConversationListFragment` both need the locale and it is my understanding that using constructor arguments for fragments is poor form. The interface exists so that the fragments can get a reference to Locale `onAttach()`.

A protected `getLocale()` method could be added to both the activities and then their fragments could grab it `onAttach()` but casting activities like that also seemed kinda poor form.

**Comment by moxie0** [comment]
locale is serializable

**Comment by McLoo** [comment]
for the record:
 In 2.9.0 times (minutes, hours ) change between TS and system locale. Dates still show in system locale. 
Didn't figure out the when they switch. Will recheck, ones this is merged.

File: src/org/thoughtcrime/securesms/ConversationActivity.java [hunk 3]
```diff
@@ -725,6 +727,8 @@ private void initializeResources() {
     masterSecret     = getIntent().getParcelableExtra(MASTER_SECRET_EXTRA);
 
     recipients.addListener(this);
+
+    getIntent().putExtra(DYNAMIC_LOCALE_EXTRA, dynamicLanguage.getCurrentLocale());
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/ConversationActivity.java]: look at mcginty's routingactivity pr and follow his lead

File: src/org/thoughtcrime/securesms/ConversationFragment.java [hunk 4]
```diff
@@ -61,6 +61,7 @@
   private Recipients   recipients;
   private long         threadId;
   private ActionMode   actionMode;
+  private Locale       dynamicLocale;
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/ConversationFragment.java]: what's dynamic about it?

File: src/org/thoughtcrime/securesms/ConversationFragment.java [hunk 5]
```diff
@@ -102,14 +103,17 @@ public void onNewIntent() {
   }
 
   private void initializeResources() {
-    this.masterSecret = this.getActivity().getIntent().getParcelableExtra("master_secret");
-    this.recipients   = RecipientFactory.getRecipientsForIds(getActivity(), getActivity().getIntent().getLongArrayExtra("recipients"), true);
-    this.threadId     = this.getActivity().getIntent().getLongExtra("thread_id", -1);
+    final Intent parentIntent = this.getActivity().getIntent();
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/ConversationFragment.java]: unnecessary

File: src/org/thoughtcrime/securesms/util/DynamicLanguage.java [hunk 6]
```diff
@@ -35,6 +35,10 @@ public void updateServiceLocale(Service service) {
     setContextLocale(service, currentLocale);
   }
 
+  public Locale getCurrentLocale() {
+    return (currentLocale != null) ? currentLocale : Locale.getDefault();
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/util/DynamicLanguage.java]: when would you want this to be true?

File: src/org/thoughtcrime/securesms/PassphraseRequiredActionBarActivity.java [hunk 7]
```diff
@@ -78,6 +82,20 @@ public void onMasterSecretCleared() {
...
+                                                @NonNull Locale locale) {
+    Bundle args = new Bundle();
+    args.putParcelable("master_secret", masterSecret);
+    args.putSerializable(ARG_LOCALE, locale);
+    fragment.setArguments(args);
+    getSupportFragmentManager().beginTransaction()
+                               .replace(target, fragment)
+                               .commit();
+    return fragment;
+  }
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/PassphraseRequiredActionBarActivity.java]: one of these methods should call the other

File: src/org/thoughtcrime/securesms/util/DateUtils.java [hunk 8]
```diff
@@ -49,34 +55,34 @@ public static String getBriefRelativeTimeSpanString(final Context c, final long
...
-      if (isWithin(timestamp, 6, TimeUnit.DAYS)) {
-        formatFlags |= DateUtils.FORMAT_SHOW_WEEKDAY | DateUtils.FORMAT_ABBREV_WEEKDAY;
-      } else if (isWithin(timestamp, 365, TimeUnit.DAYS)) {
-        formatFlags |= DateUtils.FORMAT_SHOW_DATE | DateUtils.FORMAT_NO_YEAR | DateUtils.FORMAT_ABBREV_ALL;
-      } else {
-        formatFlags |= DateUtils.FORMAT_SHOW_DATE | DateUtils.FORMAT_ABBREV_ALL;
-      }
-      return DateUtils.formatDateTime(c, timestamp, formatFlags);
+      String format = "";
+      if      (isWithin(timestamp,   6, TimeUnit.DAYS)) format += "EEE ";
```
Comment (mcginty) [inline src/org/thoughtcrime/securesms/util/DateUtils.java]: the += doesn't make sense here since format is always empty

Comment (rhodey) [inline src/org/thoughtcrime/securesms/util/DateUtils.java]: good catch, I'll try the StringBuilder pattern and see how that works out.

File: src/org/thoughtcrime/securesms/PassphraseRequiredActionBarActivity.java [hunk 9]
```diff
@@ -78,9 +82,12 @@ public void onMasterSecretCleared() {
 
   protected <T extends Fragment> T initFragment(@IdRes int target,
                                                 @NonNull T fragment,
-                                                @NonNull MasterSecret masterSecret) {
+                                                @NonNull MasterSecret masterSecret,
+                                                Locale locale) {
```
Comment (mcginty) [inline src/org/thoughtcrime/securesms/PassphraseRequiredActionBarActivity.java]: `@Nullable`

File: src/org/thoughtcrime/securesms/PassphraseRequiredActionBarActivity.java [hunk 10]
```diff
@@ -78,9 +82,12 @@ public void onMasterSecretCleared() {
...
     Bundle args = new Bundle();
     args.putParcelable("master_secret", masterSecret);
+    args.putSerializable(ARG_LOCALE, locale);
```
Comment (mcginty) [inline src/org/thoughtcrime/securesms/PassphraseRequiredActionBarActivity.java]: `LOCALE_EXTRA` i think is the canonical form for bundle constants

**Pull request closed without merging**